### PR TITLE
Add P0443r14 executor and a a few P1897 algorithms

### PR DESCRIPTION
**Pull request by msimberg**
Adds the remaining CPOs and traits from P0443r14 (see #5045). I've used the fallback implementations from P0443r14 pretty much verbatim, so there might be room for changes here.

Also adds `transform` and `sync_wait` as the first algorithms (and `wait`, `get` for symmetry with the current `future` functionality, although these should proably not stay). Also adds a `sink_receiver` (should be called `void_receiver`?).

Adds a simple executor for our coroutines based on the new functionality. The `bulk_execute` functionality is incomplete but I'd like to keep it like this for now until the proposal evolves a bit more.

Extends `tag_fallback_invoke` with `tag_priority_invoke`, which gives three options: `tag_override_invoke` which always takes precedence over user-defined `tag_invoke`s, then uses user-defined `tag_invoke`s, and finally uses `tag_fallback_invoke` if nothing else matches.

@K-ballo (and others interested), I had a need for a type list `transform` and `unique`. I'm sure you have a better idea of how to do that efficiently (especially `unique`), so I'd appreciate your eyes on that (see https://github.com/msimberg/hpx/blob/1fa6ec165d410d0508183cb0e1d532e1840f0cf6/libs/core/type_support/include/hpx/type_support/pack.hpp#L196-L237).

Includes #5033.

To do:
- [x] Add documentation for the new CPOs
- [x] Update documentation for the existing CPOs
- [x] Move traits out from `traits` namespace (objections?)

### CODE REVIEW

**Comment by msimberg** [comment]
retest

**Comment by msimberg** [comment]
I welcome reviews on this one.

File: libs/core/functional/include/hpx/functional/tag_priority_invoke.hpp [hunk 1]
```diff
@@ -0,0 +1,301 @@
...
+        constexpr HPX_FORCEINLINE auto operator()(Args&&... args) const noexcept
+            -> tag_fallback_invoke_result_t<Tag, Args&&...>
+        {
+            return hpx::functional::tag_fallback_invoke(
+                static_cast<Tag const&>(*this), std::forward<Args>(args)...);
+        }
+    };
+}}    // namespace hpx::functional
+
+#endif
```
Comment (hkaiser) [inline libs/core/functional/include/hpx/functional/tag_priority_invoke.hpp:301]: I like this!

**Review by hkaiser — DISMISSED** [review]
LGTM, thanks!

**Comment by msimberg** [comment]
> LGTM, thanks!

Thanks @hkaiser for looking through this! I'm turning this back into a draft because it looks like gcc 7 and 8 are not happy with the changes. I'm not planning on making any functional changes.

**Comment by hkaiser** [comment]
One question I had was whether we should use `tag_fallback_invoke` instead of `tag_invoke` where possible (based on our experience with segregating the segmented algorithms).

**Comment by msimberg** [comment]
> One question I had was whether we should use `tag_fallback_invoke` instead of `tag_invoke` where possible (based on our experience with segregating the segmented algorithms).

Absolutely! The P0443 customization points use `tag_priority` which gives both `tag_fallback_invoke` and `tag_priority_invoke` (the latter because according to P0443 member functions should always take precedence over free functions, or tag_invoke overloads in our case). `sync_wait` and `transform` are not even CPOs here, but should be, and should use `tag_fallback_invoke` (I've already changed this in upcoming commits).

**Comment by msimberg** [comment]
This *should* now be ok. Let's see if CI agrees. GCC 7 and 8 seemed to be more eager in instantiating templates compared to newer versions of GCC and clang.

**Pull request merged**